### ISSUE TITLE
DNS resolution fails after activating an exit node

### ISSUE THREAD

**Issue by DentonGentry** [tailscale/tailscale#3285]
### What is the issue?

After activating an exit node: `tailscale up --exit-node=100.77.180.119 --exit-node-allow-lan-access`

DNS begins failing:
```
Nov 10 04:45:50 node-red tailscaled[228029]: magicsock: endpoints changed: 24.4.x.y:41647 (portmap), [2602:248::z]:41641 (stun), 10.1.10.28:41641 (local)
Nov 10 04:45:50 node-red tailscaled[228029]: control: client.newEndpoints(0, [24.4.x.y:41647 [2602:248::z:41641 10.1.10.28:41641])
Nov 10 04:45:50 node-red tailscaled[228029]: control: mapRoutine: state:authenticated
Nov 10 04:45:53 node-red tailscaled[228029]: control: mapRoutine: state:authenticated
Nov 10 04:45:53 node-red tailscaled[228029]: allowing exit node access to internal IPs: [127.0.0.0/8]
Nov 10 04:45:53 node-red tailscaled[228029]: wgengine: Reconfig: configuring userspace wireguard config (with 2/26 peers)
Nov 10 04:45:53 node-red tailscaled[228029]: wgengine: Reconfig: configuring router
Nov 10 04:45:53 node-red tailscaled[228029]: wgengine: Reconfig: configuring DNS
Nov 10 04:45:53 node-red tailscaled[228029]: dns: Set: {DefaultResolvers:[[2001:4860:4860::8844]:53 [2001:4860:4860::8888]:53 8.8.8.8:53 8.8.4.4:53] Routes:{example.com.beta.tailscale.net.:[] ts-dns.test.:[100.120.55.47:53] ts.net.:[]}+65arpa SearchDomains:[chihuahua-tyrannosaurus.ts.net. example.com.beta.tailscale.net. ts-dns.test.] Hosts:52}
Nov 10 04:45:53 node-red tailscaled[228029]: dns: Resolvercfg: {Routes:{.:[[2001:4860:4860::8844]:53 [2001:4860:4860::8888]:53 8.8.8.8:53 8.8.4.4:53] ts-dns.test.:[100.120.55.47:53]} Hosts:52 LocalDomains:[example.com.beta.tailscale.net. ts.net.]+65arpa}
Nov 10 04:45:53 node-red tailscaled[228029]: dns: OScfg: {Nameservers:[100.100.100.100] SearchDomains:[chihuahua-tyrannosaurus.ts.net. example.com.beta.tailscale.net. ts-dns.test.] MatchDomains:[]}
Nov 10 04:46:15 node-red tailscaled[228029]: dns: error: context deadline exceeded
Nov 10 04:46:20 node-red tailscaled[228029]: dns: error: context deadline exceeded
Nov 10 04:46:25 node-red tailscaled[228029]: [RATELIMIT] format("dns: resolver: forward: DoH error from %v: %v") (4 dropped)
Nov 10 04:46:25 node-red tailscaled[228029]: dns: resolver: forward: DoH error from 8.8.8.8: Post "https://dns.google/dns-query": dial tcp 8.8.4.4:443: i/o timeout
```

However DNS from the shell still works, and I can ping 8.8.8.8:
```
root@node-red ~# dig @8.8.8.8 ifconfig.co
...removed...
;; ANSWER SECTION:
ifconfig.co.            255     IN      A       172.64.80.1

root@node-red ~# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=120 time=72.8 ms
root@node-red ~# 
```

### OS

Linux

### OS version

Debian 10

### Tailscale version

1.17.205

### Bug report

BUG-133a89e11eeea80b0160018b0b3ad07088704bbb22b346c9f0d203bebad6294a-20211110125050Z-2e8a6a99f2274c39

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_3tivl)

**Comment by twpayne** [comment]
I am seeing this issue on macOS (Tailscale v1.16.2) with a Linux exit node (Tailscale v1.18.0). #1535 might also be the same issue on Windows.

In my case I see exactly the same issue as the OP, namely that as soon as I use an exit node DNS resolution stops working, seemingly because Tailscale removes required DNS settings, even though traffic to the DNS server remains routable.

With no exit node, the macOS's DNS settings are correct and DNS resolution works:

```console
$ scutil --dns | head -n 8
DNS configuration

resolver #1
  nameserver[0] : 8.8.8.8
  nameserver[1] : 8.8.4.4
  if_index : 11 (en0)
  flags    : Request A records
  reach    : 0x00000002 (Reachable)
```

```console
$ scutil --dns | tail -n 8
DNS configuration (for scoped queries)

resolver #1
  nameserver[0] : 8.8.8.8
  nameserver[1] : 8.8.4.4
  if_index : 11 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00000002 (Reachable)
```

```console
$ dig tailscale.com

; <<>> DiG 9.10.6 <<>> tailscale.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 6836
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;tailscale.com.			IN	A

;; ANSWER SECTION:
tailscale.com.		42	IN	A	104.16.244.78
tailscale.com.		42	IN	A	104.16.243.78

;; Query time: 3 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sat Nov 20 21:39:39 +04 2021
;; MSG SIZE  rcvd: 63
```

When I select the exit node on the macOS machine to be the Linux exit node, then DNS lookup times out. Note that I *do* have "Allow local network access" enabled on the macOS machine:

```console
$ dig tailscale.com

; <<>> DiG 9.10.6 <<>> tailscale.com
;; global options: +cmd
;; connection timed out; no servers could be reached
```

However, the DNS server remains reachable:

```console
$ dig @8.8.8.8 tailscale.com

; <<>> DiG 9.10.6 <<>> @8.8.8.8 tailscale.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 53810
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;tailscale.com.			IN	A

;; ANSWER SECTION:
tailscale.com.		60	IN	A	104.16.244.78
tailscale.com.		60	IN	A	104.16.243.78

;; Query time: 134 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sat Nov 20 21:47:33 +04 2021
;; MSG SIZE  rcvd: 63
```

However, adding the exit node has removed the DNS server (compare this with the previous output of `scutil --dns | head -n 8` above):

```console
$ scutil --dns | head -n 8
DNS configuration

resolver #1
  flags    : Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)

resolver #2
  domain   : local
```

Interesting, the DNS configuration (for unscoped queries) remains:

```console
$ scutil --dns | tail -n 8
DNS configuration (for scoped queries)

resolver #1
  nameserver[0] : 8.8.8.8
  nameserver[1] : 8.8.4.4
  if_index : 11 (en0)
  flags    : Scoped, Request A records
  reach    : 0x00000002 (Reachable)
```

Please do tell me if you need any more information.

**Comment by twpayne** [comment]
I found that if I explicitly add the DNS serves (8.8.8.8 and 8.8.4.4 in my case) to Tailscale through the admin panel (https://login.tailscale.com/admin/dns) then DNS works after the exit node is set 🎉 . Note that I have MagicDNS disabled and  override local DNS off.

I'm not sure _why_ this works, I don't think it should be necessary, but I'm glad that it does :)

**Comment by nikandfor** [comment]
Thank @twpayne it worked for me either. And even it worked without local DNS overriding.

**Comment by DentonGentry** [comment]
1.19.x builds in https://pkgs.tailscale.com/unstable/ include functionality to forward DNS queries to the exit node, which is expected to resolve most issues like this. This functionality will be in the 1.20 release build.

Both the Windows client and the exit node would need to be running 1.19.x or later.

**Comment by twpayne** [comment]
Awesome, thank you Denton!

**Comment by mayakacz** [comment]
Is this fixed with 1.20?

**Comment by DentonGentry** [comment]
Fixed in 1.20
(there were some followon issues with this first exit node DNS feature, tracked in https://github.com/tailscale/tailscale/issues/3842 and expected to be fixed in 1.22)
